p2p/host/basic/basic_host.go

```diff
@@ -783,6 +828,13 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
 	return dedupAddrs(finalAddrs)
 }
 
+// TODO Do this for routed & relay hosts as well
+// RegisterProvider allows the host to be an IntrospectorRegistry. Please look at the docs for the IntrospectorRegistry
+// interface in go-libp2p-core for more details.
+func (h *BasicHost) RegisterProviders(p *coreit.ProvidersTree) error {
```

Comment:
I wouldn't add this on the `BasicHost` directly, it muddies the namespace and the term "Provider" is not contextualised. Let's add another layer of indirection in between.

```go
package host

type IntrospectableHost interface {
    Introspector() inspect.Introspector
}
```

Then a consumer can do:

```go
if ins, ok := h.(host.IntrospectableHost); ok {
    ins.Introspector().FetchCurrentState()
}
```

Reply:
I see what you mean. In this case, we can remove the `IntrospectorRegistry` interface defined [here](https://github.com/libp2p/go-libp2p-core/blob/d0dd06b3894f427b8438ea21c73bd724220d565b/introspect/introspect.go#L7) as modules don't need to check if a host is a `Registry` or not anymore. They can simply check if it has an `Introspector`.

This would have the side effect of exposing the `FetchCurrentState` to modules whose primary job is to register providers but I think we can live with that.

Reply:
Done